core/src/sumeragi/main_loop.rs

```diff
@@ -1113,70 +1101,68 @@ enum BlockSyncError {
 
 fn handle_block_sync(
     block: SignedBlock,
-    wsv: &WorldStateView,
-    finalized_wsv: &WorldStateView,
+    state: &State,
 ) -> Result<BlockSyncOk, (SignedBlock, BlockSyncError)> {
     let block_height = block.payload().header.height;
-    let wsv_height = wsv.height();
-    if wsv_height + 1 == block_height {
+    let state_height = state.view().height();
+    if state_height + 1 == block_height {
         // Normal branch for adding new block on top of current
-        let mut new_wsv = wsv.clone();
+        let mut state_block = state.block(false);
         let topology = {
-            let last_committed_block = new_wsv
+            let last_committed_block = state_block
                 .latest_block_ref()
                 .expect("Not in genesis round so must have at least genesis block");
-            let new_peers = new_wsv.peers_ids().clone();
+            let new_peers = state_block.world.peers_ids().clone();
             let view_change_index = block.payload().header().view_change_index;
             Topology::recreate_topology(&last_committed_block, view_change_index, new_peers)
         };
-        ValidBlock::validate(block, &topology, &mut new_wsv)
+        ValidBlock::validate(block, &topology, &mut state_block)
             .and_then(|block| {
                 block
                     .commit(&topology)
                     .map_err(|(block, err)| (block.into(), err))
             })
-            .map(|block| BlockSyncOk::CommitBlock(block, new_wsv))
+            .map(|block| BlockSyncOk::CommitBlock(block, state_block))
             .map_err(|(block, error)| (block, BlockSyncError::BlockNotValid(error)))
-    } else if wsv_height == block_height && block_height > 1 {
+    } else if state_height == block_height && block_height > 1 {
         // Soft-fork on genesis block isn't possible
         // Soft fork branch for replacing current block with valid one
-        let mut new_wsv = finalized_wsv.clone();
+
+        let peer_view_change_index = state.view().latest_block_view_change_index();
+        let block_view_change_index = block.payload().header.view_change_index;
+        if peer_view_change_index >= block_view_change_index {
+            return Err((
+                block,
+                BlockSyncError::SoftForkBlockSmallViewChangeIndex {
+                    peer_view_change_index,
+                    block_view_change_index,
+                },
+            ));
+        }
+
+        let mut state_block = state.block(true);
```

Comment:
We need 2 levels of atomicity:
- block level: this is required not only for soft-forks but as well for cases when we just drop block without committing it
- transaction level: for triggers and transactions rollbacks

Reply:
This may be a terminology issue, but
I think I was wrong to call this line "block-level rollback".
I might have been better to call it "revert" instead of "rollback", as it seems to revert a block that has already been committed.
The same could be said for the `rollback_latest_block` params and the `rollback` fields of the `storage` crate structures.

And a feature that is a "revert" and not a "rollback" in this sense might be not a place to use `concread`, which was the intent of my original question

Reply:
Ok, let's consider all possible execution paths for state to eliminate terminology issue.

1.a) new block changes are created on top of existing state (normal mode)
1.b) latest block changes are reverted and then new block changes are created on top reverted state (revert/rollback/soft-fork mode)
2) new transaction changes are created
2.a) either transaction is explicitly applied (apply)
2.b) otherwise transaction changes are implicitly discarded (abort)
3.a) either block changes are explicitly committed (commit)
3.b) otherwise block changes are discarded (abort)

Reply:
Thank you for clarification, it is okay to me except for 1.b) I'd call it revert instead of rollback to go back beyond a commit point

Reply:
Why two `..WriteTxn` fields (`rollback` and another) are necessary in the `storage` crate structures, even though a single `..WriteTxn` has the transactional functionality (either commit or rollback)?
I guess it is due to the requirement of the two levels of atomicity

Reply:
I still don't understand the situation that requires block-level atomicity as a new feature.
At transaction-level it makes sense because instruction executions can fail in the halfway state.
On the other hand, a block effect on the state, I think, has been all or nothing even without the new atomic mechanism

Reply:
>  I'd call it revert instead of rollback to go back beyond a commit point

Good point, i will adjust naming.

> I guess it is due to the requirement of the two levels of atomicity

Ah, we need `rollback` field to be transactional in case where transaction is dropped we need to restore previous `rollback` version.
So we need this to revert latest block (soft-fork handling).

> I think, has been all or nothing even without any new atomic mechanism

It's not new mechanism, currently to handle soft-forks we have WSV one block behind called `finalized_wsv` so in case of soft-fork it replaces original one.
This revert is just more smart way around it.

Reply:
Renamed `rollback` to `revert`. 
Also splitted `block(revert_previous_block)` into 2 methods `block()` and `block_and_revert()`.